src/OpenToolkit.Mathematics/Color/Color3.cs

```diff
@@ -0,0 +1,1231 @@
+﻿using System;
+using System.Diagnostics.Contracts;
+using OpenToolkit.Mathematics;
+
+namespace OpenToolkit
+{
+    /// <summary>
+    /// Interface for ColorSpaces with 3 elements.
+    /// </summary>
+    public interface IColorSpace3
+    {
+    }
+
+    /// <summary>
+    /// Red, green, blue colorspace.
+    /// </summary>
+    public sealed class Rgb : IColorSpace3
+    {
+    }
+
+    /// <summary>
+    /// Hue, saturation, value colorspace.
+    /// </summary>
+    public sealed class Hsv : IColorSpace3
+    {
+    }
+
+    /// <summary>
+    /// Hue, saturation, light colorspace.
+    /// </summary>
+    public sealed class Hsl : IColorSpace3
+    {
+    }
+
+    /// <summary>
+    /// Implementation of a color3 colorspace.
+    /// </summary>
+    /// <typeparam name="T">The color space of the given color.</typeparam>
+    public struct Color3<T> : IEquatable<Color3<T>>
+        where T : IColorSpace3
+    {
+        /// <summary>
+        /// The first element of this color. This corresponds to the first letter of the color type.
+        /// </summary>
+        public float X;
+
+        /// <summary>
+        /// The second element of this colour. This corresponds to the second letter of the colour type.
+        /// </summary>
+        public float Y;
+
+        /// <summary>
+        /// The third element of this color. This corresponds to the second letter of the colour type.
+        /// </summary>
+        public float Z;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Color3{T}"/> struct.
+        /// </summary>
+        /// <param name="x">The first element of the color.</param>
+        /// <param name="y">The second element of the color.</param>
+        /// <param name="z">The third element of the color.</param>
+        public Color3(float x, float y, float z)
```

Comment:
Please beat me up, if I dont get it, but shouldnt it be r, g, b here, same for all the other color3/4s and h, s, v for ColorHsv, and so forth, while the _internal_ container for rgb, hsv, hsl and so forth is xyz. but xyz should never be exposed... semantically speaking.

Reply:
This wouldnt really make sense as you would then also have Color4<Rgba>.H, what is H now? and the x component can mean the alpha for certain color spaces like Argb. So this wouldnt make sense really, what Varon and i have discussed is making a function like Color4<Rgba>.A() which might either return or set the a component of a Color4, but since extension properties arent a thing, what you are suggested cant really be done. Sorry.

Reply:
Yes, I'm a dumdum. While writing the first comment I thought there were different ctors per colorspace.... only then it would have made sense. Its all good.